cmd/autoscaler/main.go

```diff
@@ -119,75 +116,161 @@ func main() {
+		})
+		eg.Go(statsServer.ListenAndServe)
+		eg.Go(profilingServer.ListenAndServe)
+
+		// This will block until either a signal arrives or one of the grouped functions
+		// returns an error.
+		<-egCtx.Done()
+
+		statsServer.Shutdown(5 * time.Second)
+		profilingServer.Shutdown(context.Background())
+		// Don't forward ErrServerClosed as that indicates we're already shutting down.
+		if err := eg.Wait(); err != nil && err != http.ErrServerClosed {
+			logger.Errorw("Error while running server", zap.Error(err))
+		}
 	}
 
-	// Start all of the informers and wait for them to sync.
-	if err := controller.StartInformers(ctx.Done(), informers...); err != nil {
-		logger.Fatalw("Failed to start informers", zap.Error(err))
+	recorder := controller.GetEventRecorder(ctx)
+	if recorder == nil {
+		// Create event broadcaster
+		logger.Debug("Creating event broadcaster")
+		eventBroadcaster := record.NewBroadcaster()
+		watches := []watch.Interface{
+			eventBroadcaster.StartLogging(logger.Named("event-broadcaster").Infof),
+			eventBroadcaster.StartRecordingToSink(
+				&typedcorev1.EventSinkImpl{Interface: kubeclient.Get(ctx).CoreV1().Events("")}),
+		}
+		recorder = eventBroadcaster.NewRecorder(
+			// todo: what agent name?
+			scheme.Scheme, corev1.EventSource{Component: "fake-agent-name"})
+		go func() {
+			<-ctx.Done()
+			for _, w := range watches {
+				w.Stop()
+			}
+		}()
 	}
 
-	go controller.StartAll(ctx.Done(), controllers...)
+	// Set up leader election config
+	leaderElectionConfig, err := sharedmain.GetLeaderElectionConfig(ctx)
+	if err != nil {
+		log.Fatalf("Error loading leader election configuration: %v", err)
+	}
+	leConfig := leaderElectionConfig.GetComponentConfig(component)
 
-	go func() {
-		for sm := range statsCh {
-			collector.Record(sm.Key, sm.Stat)
-			multiScaler.Poke(sm.Key, sm.Stat)
-		}
-	}()
+	if !leConfig.LeaderElect {
+		log.Printf("autoscaler will not run in leader-elected mode")
+		run(ctx)
+		panic("unreachable")
+	}
 
-	profilingServer := profiling.NewServer(profilingHandler)
+	// create a unique identifier so that two controllers on the same host don't
+	// race.
+	id, err := os.Hostname()
+	if err != nil {
+		logger.Fatalw("Failed to get hostname for leader election", zap.Error(err))
+	}
+	id = id + "_" + string(uuid.NewUUID())
+	log.Printf("%v will run in leader-elected mode with id %v", component, id)
+
+	rl, err := resourcelock.New(leConfig.ResourceLock,
+		system.Namespace(), // use namespace we are running in
+		component,          // component is used as the resource name
+		kubeclient.Get(ctx).CoreV1(),
+		kubeclient.Get(ctx).CoordinationV1(),
+		resourcelock.ResourceLockConfig{
+			Identity:      id,
+			EventRecorder: recorder,
+		})
+	if err != nil {
+		logger.Fatalw("error creating lock: %v", err)
+	}
 
-	eg, egCtx := errgroup.WithContext(ctx)
-	eg.Go(func() error {
-		return customMetricsAdapter.Run(ctx.Done())
+	leaderelection.RunOrDie(ctx, leaderelection.LeaderElectionConfig{
```

Comment:
IIUC we get all of this for free in the other binaries due to their use of sharedmain?